include/revng/Support/Undirected.h

```diff
@@ -0,0 +1,142 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/ADT/iterator.h"
+
+#include "revng/Support/Assert.h"
+
+namespace llvm {
+
+/// This class is used as a marker class to tell the graph iterator to treat the
+/// underlying graph as an undirected one
+template<class GraphType>
+struct Undirected {
+  const GraphType &Graph;
+
+  inline Undirected(const GraphType &G) : Graph(G) {}
+};
+
+} // namespace llvm
+
+/// Custom child iterator which iterates over the concatenation of the
+/// successors and predecessors of a node. This is used for the `Undirected`
+/// `GraphTraits` specialization. This is done taking inspiration from the
+/// internals of `llvm::concat` implementation, but overcomes one of its
+/// limitations. The limitation is caused by the fact that internally, the `*`
+/// operator of `llvm::concat` (and more specifically the `getHelper`), first
+/// dereferences the internal iterator that is being concatenated, and then
+/// takes and returns its pointer (`&*It`). If the iterator used internally,
+/// when dereferenced, returns a temporary object with a limited lifetime, which
+/// goes out of scope before its used by the user of `llvm::concat`, we have a
+/// problem. Instead here, we artificially extend the lifetime of the object
+/// pointed to by the internal iterator, wrapping it into the `Proxy` struct.
+/// When dereferenciation happens, no problems of lifetime happen.
+// TODO: An alternative to implementing our custom iterator, would be to provide
+//       our own implementation of `revng::concat`, which does not have the
+//       limitation described above.
```

Comment:
After today's talk, please replace this comment with "drop this once we adopt C++34 or we could adopt rangesv3" and link the specific relevant file.

Reply:
I added a comment linking the page on `cppreference` for C++26 and the documentation for range-v3

https://github.com/revng/revng/blob/bf979cc12411a1e94654b1f77e6cfb0054b6daed/include/revng/Support/Undirected.h#L37

Reply:
C++34 :heart: :rofl: